### PR TITLE
Write only the extent of ancestors to improve generation perf

### PR DESCRIPTION
**Pull request by benjeffery**
This takes the `ga-gen` step from `3:41` to `0:18`. Writing the entire ancestor was slow. Flame graph before:
Screenshot at 2023-02-21 13-46-29

And after:
Screenshot at 2023-02-21 13-46-07
The large amount of time formatting strings is due to the logging lines, but I think this os only used in the synchronous code.

### CODE REVIEW

File: tsinfer/formats.py [hunk 1]
```diff
@@ -247,8 +259,15 @@ def add(self, **kwargs):
             self._queue_flush_buffer()
         offset = self.num_buffered_items[self.write_buffer]
         for key, value in kwargs.items():
-            if "full_haplotype" in key:
-                self.buffers[key][self.write_buffer][:, offset, 0] = value
+            # Here we have to special case the haplotype for performance
+            # reasons, as writing the full haplotype is expensive.
+            if key == "haplotype":
```
Comment (jeromekelleher) [inline tsinfer/formats.py:264]: Could make this a bit more readable by pulling start and end out into local vars (rather then looking up kwargs repeatedly)

File: tsinfer/formats.py [hunk 2]
```diff
@@ -2735,7 +2755,9 @@ def num_sites(self):
         """
         The number of inference sites used to generate the ancestors
         """
-        return self.sites_position.shape[0]
+        if not hasattr(self, "_num_sites"):
```
Comment (jeromekelleher) [inline tsinfer/formats.py:2758]: Using hasattr feels a bit icky (and slow I think?). How about initialising ``self._num_sites = None`` and using that as the caching tag?

**Review by jeromekelleher — APPROVED** [review]
LGTM, a couple of minor comments

File: tsinfer/formats.py (hunk 2 above)
Comment (benjeffery) [reply tsinfer/formats.py:2758]: I wrote it that way first - but `load` calls `__new__` and not `__init__`. The best way is probably a caching decorator. Will see if our python version supports using the one in functools.

File: tsinfer/formats.py (hunk 1 above)
Comment (benjeffery) [reply tsinfer/formats.py:265]: Fixed

File: tsinfer/formats.py (hunk 2 above)
Comment (benjeffery) [reply tsinfer/formats.py]: Woop! Now that we have dropped 3.7 we can use `@functools.cached_property`!

**Pull request merged**